### DISCUSSION TITLE
How to speed up measurement initialization?

### DISCUSSION THREAD

**Discussion by moritzj29** [pymeasure/pymeasure#467] [Q&A] [answered]
Hi everyone,
I currently use pymeasure's ManagedWindow to run a couple of parameter sweeps. I add the procedures directly in my python script initializing the managed window using loops. When running the script, all the procedures and result objects are created. The files are written to the disk. After that the measurements run and I can observe the progress in the MangedWindow. So far so good.

Sometimes I do sweeps with a couple of thousand measurements, so 5000 files for example. This works, but pymeasure takes like 2-3h just for initializing the files on the disk. Only the first measurement is ran immediatley while still setting up the files. The subsequent measurements only start after all files have been set up. Also the ManagedWindow is blank/unresponsive until all measurements have been set up.

Is there a way to speed this up? Like with the first measurement which is running while the files are set up.
I tried adding sleeps so the CPU has some spare time, but this doesn't help.
Do I need to go for an additional worker?

I'm curious for your ideas!

EDIT: add example code (mainly `gui_sequencer.py` example code with nested `for` loop at the end)

```
"""
This example demonstrates how to make a graphical interface, and uses
a random number generator to simulate data so that it does not require
an instrument to use. It also demonstrates the use of the sequencer module.

Run the program by changing to the directory containing this file and calling:

python gui_sequencer.py

"""

import sys
sys.modules['cloudpickle'] = None
import random
import tempfile
from time import sleep

import logging
log = logging.getLogger('')
log.addHandler(logging.NullHandler())

from pymeasure.experiment import Procedure, IntegerParameter, Parameter, \
    FloatParameter
from pymeasure.experiment import Results
from pymeasure.display.Qt import QtGui
from pymeasure.display.windows import ManagedWindow
from pymeasure.display.widgets import SequencerWidget


class TestProcedure(Procedure):

    iterations = IntegerParameter('Loop Iterations', default=100)
    delay = FloatParameter('Delay Time', units='s', default=0.2)
    seed = Parameter('Random Seed', default='12345')

    DATA_COLUMNS = ['Iteration', 'Random Number']

    def startup(self):
        log.info("Setting up random number generator")
        random.seed(self.seed)

    def execute(self):
        log.info("Starting to generate numbers")
        for i in range(self.iterations):
            data = {
                'Iteration': i,
                'Random Number': random.random()
            }
            log.debug("Produced numbers: %s" % data)
            self.emit('results', data)
            self.emit('progress', 100 * i / self.iterations)
            sleep(self.delay)
            if self.should_stop():
                log.warning("Catch stop command in procedure")
                break

    def shutdown(self):
        log.info("Finished")


class MainWindow(ManagedWindow):

    def __init__(self):
        super(MainWindow, self).__init__(
            procedure_class=TestProcedure,
            inputs=['iterations', 'delay', 'seed'],
            displays=['iterations', 'delay', 'seed'],
            x_axis='Iteration',
            y_axis='Random Number',
            sequencer=True,
            sequencer_inputs=['iterations', 'delay', 'seed'],
            # sequence_file="gui_sequencer_example_sequence.txt",
            inputs_in_scrollarea=True
        )
        self.setWindowTitle('GUI Example')

    def queue(self, *, procedure=None):
        filename = tempfile.mktemp()

        if procedure is None:
            procedure = self.make_procedure()

        results = Results(procedure, filename)
        experiment = self.new_experiment(results)

        self.manager.queue(experiment)


if __name__ == "__main__":
    app = QtGui.QApplication(sys.argv)
    window = MainWindow()
    window.show()

    for i in range(0, 100):
        for j in range(0, 100):
            QtGui.QApplication.processEvents()  # solution, as proposed by @CasperSchippers 
            window.queue()

    sys.exit(app.exec_())
```

**Comment by bilderbuchi** [comment]
That sounds really weird. Getting a Profile would be  useful to see where all that time is spent. Failing that, can you create a minimum example with a fake instrument that shows this behaviour?  @CasperSchippers any ideas?

**Comment by CasperSchippers** [comment, marked as answer]
That is indeed pretty weird. Could you tell us how you precisely queue all the measurements? If I try to reproduce this using the SequencerWidget, I experience no problems of this kind.

I think, if you add an ordinary sleep, the CPU will not have a hard time indeed, but it does not allow for processing of events; I think if you replace the `sleep` (I presume it is placed in the loop that queues the measurements) with `QtGui.QApplication.processEvents()`, event can actually be processed and the program should not hang; I think (though I'm not entirely sure) that this will also help with starting the second measurement while the queueing is still going on.

**Reply by moritzj29** [reply]
great! that's the magic line is was missing! I also tried your suggestion with the sequencer which also worked better. It's still not super fluent, but thats probably due to old hardware... Now that the measurements run while the the remaining ones are added to the queue, I don't care how long that takes (~about 140 files/min on my slow lab-notebook without ssd...).
